firefox_puppeteer/ui/toolbars.py

```diff
@@ -0,0 +1,337 @@
+
+        Wait(self.marionette).until(
+            lambda _: self.urlbar.get_attribute('focused') == 'true')
+
+
+class AutocompleteResults(BaseLib):
+    """Library for interacting with autocomplete results.
+    """
+
+    def __init__(self, *args, **kwargs):
+        BaseLib.__init__(self, *args, **kwargs)
+        # TODO: A "utility" module that sets up the client directly would be
+        # useful here.
+        self.l10n = L10n(self.get_marionette)
+        self.keys = Keys(self.get_marionette)
+
+    @property
+    def popup(self):
+        """Returns a list of auto complete result elements.
+        """
+        return self.marionette.find_element(By.ID,
+                                            'PopupAutoCompleteRichResult')
+
+    @property
+    def results(self):
+        """Returns the container node for autocomplete results.
+        """
+        return self.popup.find_element('anon attribute',
+                                       {'anonid': 'richlistbox'})
+
+    def is_open(self):
+        return self.popup.get_attribute('state') == 'open'
+
+    def get_visible_results(self):
+        """Returns a list of visible result nodes."""
+        return self.marionette.execute_script("""
+          let rv = [];
+          let node = arguments[0];
+          let count = node.itemCount;
+          for (let i = 0; i < count; ++i) {
+            let item = node.getItemAtIndex(i);
+            if (!item.hasAttribute("collapsed")) {
+              rv.push(item);
+            }
+          }
+          return rv;
+        """, script_args=[self.results])
+
+    def close(self, force=False):
+        if self.is_open():
+            if force:
+                self.marionette.execute_script("""
+                  arguments[0].hidePopup();
+                """, script_args=[self.popup])
+            else:
+                (self.marionette.find_element('id', 'urlbar')
+                                .send_keys(Keys.ESCAPE))
+            Wait(self.marionette).until(
+                lambda mn: not self.is_open())
+
+    def get_matching_text(self, result, match_type):
+        """Retuns an array of strings of the matching text within a autocomplete
+        result in the urlbar.
+
+        :param result: The result to inspect for matches.
+        :param result_type: The type of match to search for (one of "title", "url").
+        """
+
+        assert match_type in ('title', 'url'), ('match_type provided must be one of'
+                                                '"title" or "url", not %s' % match_type)
+
+        if match_type == "title":
+            descnode = self.marionette.execute_script("""
+              return arguments[0].boxObject.firstChild.childNodes[1].childNodes[0];
+            """, script_args=[result])
+        if match_type == "url":
+            descnode = self.marionette.execute_script("""
+              return arguments[0].boxObject.lastChild.childNodes[2].childNodes[0];
+            """, script_args=[result])
```

Comment:
There is only a single difference in the index. It would be better to also push the index into the execute_script() call, and have only one such thing.